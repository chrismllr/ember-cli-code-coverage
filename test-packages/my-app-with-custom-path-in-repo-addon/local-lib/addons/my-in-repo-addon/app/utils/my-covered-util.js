export { default } from 'my-app-in-repo-addon/utils/my-covered-util';
