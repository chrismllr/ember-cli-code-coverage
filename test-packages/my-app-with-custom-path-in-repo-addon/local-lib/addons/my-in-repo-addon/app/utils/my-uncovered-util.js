export { default } from 'my-app-in-repo-addon/utils/my-uncovered-util';
