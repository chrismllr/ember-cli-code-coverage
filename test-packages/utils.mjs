import { readJSON } from 'fs-extra';
import { expect } from 'vitest';
import { Project } from 'fixturify-project';
import { exists } from 'fs-extra';
import { readFile } from 'fs/promises';
import { execa } from 'execa';

export default async function setupTestDir(APP_DIR, env, deps) {
  const project = Project.fromDir(`test-packages/${APP_DIR}`, { linkDevDeps: true });

  for (const [key,value] of Object.entries(deps)) {
    project.addDevDependency(key, value);
  }

  if (depsReinstallNeeded(deps)) {
    project.addDevDependency(`ember-cli-code-coverage`, `file:${process.cwd()}/packages/ember-cli-code-coverage`);
  }

  await project.write();

  if (deps && Object.keys(deps).length > 0) {
    await execa('pnpm', ['install', '--no-frozen-lockfile'], { cwd: project.baseDir, env });
  }

  return project.baseDir;
}
  
function depsReinstallNeeded(deps) {
  return deps && Object.keys(deps).length > 0;
}

export async function assertCoverageExists(buildPath) {
  await assertFileIsNotEmpty(`${buildPath}/lcov-report/index.html`);
  await assertFileIsNotEmpty(`${buildPath}/index.html`);
  let summary = await readJSON(`${buildPath}/coverage-summary.json`);
  expect(summary).toMatchSnapshot();
}  


export async function assertFileExists(path) {
  if (!await exists(path)) {
    throw new Error(`File ${path} does not exist`);
  }
}

export async function assertDirDoesNotExists(dir) {
  if (await exists(dir)) {
    throw new Error(`Dir ${dir} exists`);
  }
}

export async function assertFileIsNotEmpty(path) {
  return  (await readFile(path, 'utf8').length) > 0;
}