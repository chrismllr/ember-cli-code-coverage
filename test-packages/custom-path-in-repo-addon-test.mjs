'use strict';

import { execa } from 'execa';
import { assertCoverageExists, default as setupTestDir } from './utils.mjs';
import { it, describe } from 'vitest';

const BASE_PATH = 'my-app-with-custom-path-in-repo-addon';

describe('alternate in-repo addon coverage generation', function () {
  it('runs coverage on in-repo addons from a non-standard directory structure', async function () {
    let env = { COVERAGE: 'true' };
    const buildPath = await setupTestDir(BASE_PATH, env, {});
    await execa('pnpm', ['ember' , 'test', '--test-port=0'], { cwd: buildPath, env });
    await assertCoverageExists(`${buildPath}/coverage`);
  });
});
