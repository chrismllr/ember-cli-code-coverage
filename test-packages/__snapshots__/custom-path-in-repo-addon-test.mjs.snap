// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`alternate in-repo addon coverage generation > runs coverage on in-repo addons from a non-standard directory structure 1`] = `
{
  "app/app.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/router.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/utils/my-covered-util-app.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "app/utils/my-uncovered-util-app.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "statements": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
  },
  "local-lib/addons/my-in-repo-addon/addon-test-support/uncovered-test-support.js": {
    "branches": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 2,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 2,
    },
    "lines": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 4,
    },
    "statements": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 4,
    },
  },
  "local-lib/addons/my-in-repo-addon/addon/utils/my-covered-util.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "local-lib/addons/my-in-repo-addon/addon/utils/my-uncovered-util.js": {
    "branches": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "statements": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
  },
  "total": {
    "branches": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 2,
    },
    "branchesTrue": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 2,
      "pct": 28.57,
      "skipped": 0,
      "total": 7,
    },
    "lines": {
      "covered": 7,
      "pct": 46.66,
      "skipped": 0,
      "total": 15,
    },
    "statements": {
      "covered": 7,
      "pct": 46.66,
      "skipped": 0,
      "total": 15,
    },
  },
}
`;
